XMonad/Actions/MessageFeedback.hs

```diff
@@ -63,37 +71,79 @@ import Control.Applicative ((<$>))
 send :: Message a => a -> X Bool
 send = sendSM . sm
 
+-- | Behaves like 'XMonad.Operations.sendMessageWithNoRefresh', but returns True of the
+-- message was handled by the layout, False otherwise.
+sendWithNoRefresh :: Message a => a -> X Bool
+sendWithNoRefresh = sendSMWithNoRefresh . sm
+
 -- | Sends the first message, and if it was not handled, sends the second.
 -- Returns True if either message was handled, False otherwise.
 tryMessage :: (Message a, Message b) => a -> b -> X Bool
 tryMessage m1 m2 = do b <- send m1
                       if b then return True else send m2
 
+-- | Like 'tryMessage', but discards the response.
 tryMessage_ :: (Message a, Message b) => a -> b -> X ()
 tryMessage_ m1 m2 = tryMessage m1 m2 >> return ()
 
+-- | Like 'tryMessage', but without a refresh at the end.
+tryMessageWithNoRefresh :: (Message a, Message b) => a -> b -> X Bool
+tryMessageWithNoRefresh m1 m2 = do b <- sendWithNoRefresh m1
+                                   if b then return True else sendWithNoRefresh m2
+
+-- | Like 'tryMessageWithNoRefresh', but discards the response.
+tryMessageWithNoRefresh_ :: (Message a, Message b) => a -> b -> X ()
+tryMessageWithNoRefresh_ m1 m2 = tryMessageWithNoRefresh m1 m2 >> return ()
+
 -- | Tries sending every message of the list in order until one of them
 -- is handled. Returns True if one of the messages was handled, False otherwise.
 tryInOrder :: [SomeMessage] -> X Bool
 tryInOrder [] = return False
 tryInOrder (m:ms) = do b <- sendSM m
                        if b then return True else tryInOrder ms
 
+-- | Like 'tryInOrder', but discards the response.
 tryInOrder_ :: [SomeMessage] -> X ()
 tryInOrder_ ms = tryInOrder ms >> return ()
 
+-- | Like 'tryInOrder', but without a refresh at the end.
+tryInOrderWithNoRefresh :: [SomeMessage] -> X Bool
+tryInOrderWithNoRefresh [] = return False
+tryInOrderWithNoRefresh (m:ms) = do b <- sendSM m
+                                    if b then return True else tryInOrder ms
+
+-- | Like 'tryInOrderWithNoRefresh', but discards the response.
+tryInOrderWithNoRefresh_ :: [SomeMessage] -> X ()
+tryInOrderWithNoRefresh_ ms = tryInOrder ms >> return ()
+
 
 -- | Convenience shorthand for 'XMonad.Core.SomeMessage'.
 sm :: Message a => a -> SomeMessage
 sm = SomeMessage
 
 
+-- | Send a 'SomeMessage' to the corrent layout and returns true if the message was handled.
 sendSM :: SomeMessage -> X Bool
 sendSM m = do w <- workspace . current <$> gets windowset
               ml' <- handleMessage (layout w) m `catchX` return Nothing
-              updateLayout (tag w) ml'
```

Comment:
Maybe I'm misunderstanding something --- why isn't `updateLayout` called any more?

Reply:
@byorgey, @ankaan doing an update (especially with refresh) from a message handler potentially can lead to infinite loops, some care is needed.

Reply:
"update" here includes `X.O.windows` which will do a `refresh` and run `logHook` and potentially `manageHook` if there are new windows waiting to be managed.

Reply:
According to the old documentation, a refresh is expected. But that is not reflected in the actual old implementation.
The risk here is that some user has configured one of their hooks to use i.e. sendSM which could potentially trigger an infinite loop. By trial and error they would have discovered that this worked before the change. But what is actually the correct behaviour?

Reply:
Having to manually add refresh even on simple actions makes it harder to use the MessageFeedback module though. A refresh after each action is generally what is expected, which is reflected in the name of the methods. The long bulky name is the less used non-refresh variant. (As in xmonad core.)

The non-refresh is still there though if you need it and in that case you probably know you need it. It is harder to understand that you manually need to refresh after a simple action which when sending a normal message without feeback just works.